server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java

```diff
@@ -194,7 +198,12 @@ public ClusterState execute(ClusterState currentState, List<TaskContext<UpdateDe
 
             desiredNodes = DesiredNodes.updateDesiredNodesStatusIfNeeded(currentState.nodes(), desiredNodes);
 
-            return desiredNodes == initialDesiredNodes ? currentState : replaceDesiredNodes(currentState, desiredNodes);
+            if (desiredNodes == initialDesiredNodes) {
+                return currentState;
+            } else {
+                final var clusterStateWithUpdatedDesiredNodes = replaceDesiredNodes(currentState, desiredNodes);
+                return allocationService.adaptAutoExpandReplicas(clusterStateWithUpdatedDesiredNodes);
```

Comment:
nit: I would prefer:

```suggestion
                final ClusterState withUpdatedDesiredNodes = replaceDesiredNodes(currentState, desiredNodes);
                return allocationService.adaptAutoExpandReplicas(withUpdatedDesiredNodes);
```